keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+
+## Design Details
+
+### Configuration file
+
+The main part of this proposal is a configuration file. It contains an array of providers:
+
+```go
+type OIDCConfiguration struct {
+    metav1.TypeMeta
+    // Providers is a list of OIDC providers to authenticate Kubernetes users.
+    Providers []Provider
+}
+```
+
+Each provider has several properties that will be described in detail below.
+
+```go
+type Provider struct {
+    // Issuer is a basic OIDC provider connection options.
+    Issuer Issuer
+    // ClaimValidationRules are rules that are applied to validate ID token claims to authorize users.
+    // +optional
+    ClaimValidationRules []ClaimValidationRule
+    // ClaimMappings points claims of an ID token to be treated as user attributes.
+    // All mappings are logical expressions that is written in CEL https://github.com/google/cel-go.
+    ClaimMappings UserAttributes
+    // ClaimsFilter allows unmarshalling only required claims which positively affects performance.
+    // +optional
+    ClaimsFilter []string
+}
+```
+
+1. `Issuer` - is a sections for external provider specific settings, e.g., OIDC discovery URL. 
+  There is also default validation settings (ClientID, SkipOIDCValidations) that's available out of the box for flag-based 
+  OIDC provider, but for structured config they can be disabled.
+
+    ```go
+    type Issuer struct {
+        // URL points to the issuer URL in a format schema://url/path.
+        URL string
+        // CertificateAuthorityData contains PEM-encoded certificate authority certificates. Overrides CertificateAuthority
+        CertificateAuthorityData []byte
+        // ClientID the JWT must be issued for, the "sub" field. This plugin only trusts a single
+        // client to ensure the plugin can be used with public providers.
+        // Do not affect anything with the SkipOIDCValidations option enabled.
+        // +optional
+        ClientID string
+        // SkipOIDCValidations is a flag to turn off issuer validation, client id validation.
+        // OIDC related checks.
+        // +optional
+        SkipOIDCValidations bool
+   }
+   ```
+
+2. `ClaimValidationRules` - additional rules for authorization.
+    ```go
+    type ClaimValidationRule struct {
+        // Rule is a logical expression that is written in CEL https://github.com/google/cel-go.
+        Rule string
+        // Message customize returning message for validation error of the particular rule.
+        // +optional
+        Message string
+    }
+    ```
+
+    For validation expressions, the CEL is used. They are similar to validations functions for [Custom Resources](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#resource-use-by-validation-functions).
+    `Rule` expression should always evaluate a boolean. Token `claims` are passed to CEL expressions as a dynamic map `decls.NewMapType(decls.String, decls.Dyn)`.
```

Comment:
is `claims` the only variable available to the expressions? what additional macros (if any) are available to the expressions? this needs to explicitly enumerate the surface area being exposed since that becomes part of the compatibility the feature is required to maintain

Reply:
since this is using dynamic evaluation, what is the behavior if an expression references something that fails at runtime? internal error (5xx)? authentication error (401)? token is ignored? is the runtime error message exposed to the user? in the API server log?

Reply:
> is claims the only variable available to the expressions? what additional macros (if any) are available to the expressions? this needs to explicitly enumerate the surface area being exposed since that becomes part of the compatibility the feature is required to maintain

Two variables: `claim` for claims and `now` for the current time in UNIX format to be able to customize the expiration validation.

Two [extensions](https://github.com/google/cel-go/tree/master/ext#extensions): strings and encoders. 


> since this is using dynamic evaluation, what is the behavior if an expression references something that fails at runtime? internal error (5xx)? authentication error (401)? token is ignored? is the runtime error message exposed to the user? in the API server log?

500 and a log entry

I will clarify both.